crowbar_framework/app/controllers/api/restart_management_controller.rb

```diff
@@ -0,0 +1,147 @@
+  param :barclamp, String, desc: "Barclamp to apply to", required: true
+  api_version "2.0"
+
+  api :GET, "/api/restart_management/configuration", "List the disallow restart value for barclamps"
+  header "Accept", "application/vnd.crowbar.v2.0+json", required: true
+  api_version "2.0"
+  def configuration
+    if request.post?
+      configuration_post
+    else
+      configuration_get
+    end
+  end
+
+  api :POST, "/api/restart_management/restarts", "Clean the service restart flags"
+  header "Accept", "application/vnd.crowbar.v2.0+json", required: true
+  param :node, String, desc: "Node name", required: true
+  param :service, String, desc: "Service to clean restart flag for", required: true
+  api_version "2.0"
+
+  api :GET, "/api/restart_management/restarts", "Get a list of services that need restart"
+  header "Accept", "application/vnd.crowbar.v2.0+json", required: true
+  api_version "2.0"
+  def restarts
+    if request.post?
+      restarts_post
+    else
+      restarts_get
+    end
+  end
+
+  private
+
+  # Gets a list of the barclamps and extract the values from their nodes
+  # this ignores the roles and groups together all the nodes under a barclamp
+  # this is because we use a barclamp-level service restart checks instead of
+  # doing it service level, which is more cumbersome and requires us to know a list of services
+  # beforehand
+  def restarts_get
+    # maybe we want to blacklist some barclamps from the restart list?
+    blacklist = []
+    barclamps = BarclampCatalog.members("openstack").keys.reject { |m| blacklist.include? m }
+    # Chef query search to find all nodes that have requires_restart attribute not empty
+    nodes_with_restarts = NodeObject.find("requires_restart:*")
+    @final_list = {}
+    nodes_with_restarts.each do |node|
+      @final_list[node[:fqdn]] = { alias: node.alias }
+      barclamps.each do |barclamp|
+        # skip if there is not attributes for the barclamp or the requires_restart is empty
+        # why node.node.fetch? Dont ask me, ask the chef creators why this works
+        # https://github.com/SUSE-Cloud/chef/blob/10-stable-suse/chef/lib/chef/node.rb#L187
+        requires_restart = node.node.fetch(
+          "crowbar_wall", {}
+        ).fetch(
+          "requires_restart", {}
+        ).fetch(
+           barclamp.to_s, {}
+        )
+        next if requires_restart.empty?
+        @final_list[node[:fqdn]].update(barclamp => requires_restart)
+      end
+    end
+
+    render json: @final_list
+  end
+
+  def restarts_post
+    params.require(:node)
+    params.require(:service)
+
+    node_name = params[:node]
+    service = params[:service]
+    barclamps = BarclampCatalog.members("openstack").keys
+
+    # Search for the exact node with the provided params. This should yield an unique node
+    node = NodeObject.find("name:#{node_name} AND requires_restart:#{service}").first
+
+    # validate that node exists
+    raise Crowbar::Error::NotFound if node.nil?
```

Comment:
I don't think returning `NotFound` is good here, due to the query itself: the node may be found, but with no `requires_restart` flag. I'd somehow prefer if we return `NotFound` if the node doesn't exist, and something else (or even no error at all!) of there's no `requires_restart` flag for that node.

Not blocking, can be fixed in a followup PR.

Reply:
indeed, I could do a search + filtering afterwards for the node NotFound error. Im in favor of not raising an error if the flag doesnt exists.

Reply:
I modified the code to return NotFound only if the node was not found.